doc/neps/nep-0054-simd-cpp-highway.rst

```diff
@@ -0,0 +1,480 @@
+
+- get rid of the NumPy-specific templating language for more expressive C++
+- this would make using sizeless intrinsics (like for SVE) easier.
+
+In addition, we see the following considerations:
+
+- If we use Highway, we would need to switch the C++ wrappers from universal
+  intrinsics to Highway. On the other hand, the work to move to C++ is not
+  complete.
+- If we use Highway, we'd need to rewrite existing kernels using Highway
+  intrinsics. But again, moving to C++ requires touching all those kernels
+  anyway.
+- One concern regarding Highway was whether it is possible to obtain a function
+  pointer for an architecture-specific function instead of calling that
+  function directly. This so that we can be sure that calling 1-D inner loop
+  many times for a single Python API invocation does not incur the dispatching
+  overhead many times. This was investigated: this can be done with Highway
+  too.
+- A second concern was whether it's possible with Highway to allow the user at
+  runtime to select or disable dispatching to certain instruction sets. This is
+  possible.
+- Use of tags in the C++ implementation (Sayed's concern). TODO: describe in
+  more detail.
+
+
+The ``_simd`` unit testing module
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Rewriting the ``_simd testing`` module to use C++ was done very recently in PR
+`gh-24069`_. It depends on the main PR for the move to C++, `gh-21057`_.
+It allows one to access the C++ intrinsics with almost the same signature, but
+from Python. This is a great way not only for testing, but also for designing
+new SIMD kernels.
+
+It may be possible to add a similar testing and prototyping feature to Highway
+(which uses plain ``googletest``), however currently the NumPy way is quite a
+bit nicer.
+
+
+Math routines
+~~~~~~~~~~~~~
+
+Math or numerical routines are written at a higher level of abstraction than
+the universal intrinsics that are the main focus of this NEP. Highway has only
+a limited number of math routines, and they are not precise enough for NumPy's
+needs. So either way, NumPy's existing routines (which use universal
+intrinsics) will stay, and if we go the Highway route they'll simply have to
+use Highway primitives internally.
+
+There may be other libraries that have numerical routines that can be reused in
+NumPy (e.g., from SLEEF, or perhaps from JPEG XL or some other Highway-using
+libraries). There may be a small benefit here, but likely it doesn't matter too
+much.
+
+
+Supported and missing intrinsics
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Some specific intrinsics that NumPy needs may be missing from Highway, e.g.:
+
+- Supporting partial operations on both contiguous and non-contiguous memory
+  access, with both single and pair offsets. Pair offsets is crucial for
+  complex kernels that require 64-bit/32-bit stride alignment. Identified
+  intrinsics related to this:
+
+  - Highway's ``LoadInterleaved2`` would have to be extended with ``*Till``
```

Comment:
I believe this would be `MaskedLoadInterleaved2`, similar to:
https://github.com/google/highway/blob/5030a388acc3dcfada259217b9a1d387383b7a2c/hwy/ops/arm_neon-inl.h#L3225 ?
```suggestion
  - Highway's ``LoadInterleaved2`` would have to be extended with ``Masked*``
```

Reply:
Nothing to do with interleave or de-interleave, as I mentioned before in https://github.com/numpy/numpy/issues/24084#issuecomment-1621659240:


>Just to clarify, pair load/store intrinsics do not perform any interleave or de-interleave operation.
>Their purpose is to handle complex load and store operations with misaligned strides,
such as supporting a 64-bit stride on a 128-bit complex data type. We can discuss in more detail at a later time.

Using intrinsics that support mask load/store operations can have a negative impact on SIMD extensions that don't natively support these operations. It is recommended to be cautious when using such intrinsics in order to maintain optimal performance across different SIMD extensions.

Furthermore, when working with AVX512, it is advisable to minimize the usage of scatter/gather operations whenever possible. These operations tend to have high latency and throughput, which can affect performance significantly.

Reply:
I think this is straying into technical detail, I would expect Highway to present an interface that *looks correct* but may not directly map to hardware intrinsics if there's a more optimal approach.

Reply:
FYI a contributor has just [added LoadN/StoreN](https://github.com/google/highway/pull/1620).

Reply:
> FYI a contributor has just added LoadN/StoreN.

Only covers contiguous memory operation.

Reply:
I will edit this section to remove the specifics